sticker/lib/util.py

```diff
@@ -16,19 +16,104 @@
+        except Exception:
+            pass
+    else:
+        with tempfile.NamedTemporaryFile() as temp:
+            temp.write(data)
+            temp.close()
+            mime, _ = mimetypes.guess_type(temp.name)
+    return mime or "image/png"
+
+
+def video_to_gif(data: bytes, mime: str) -> bytes:
+    ext = mimetypes.guess_extension(mime)
+    if mime.startswith("video/"):
+        # run ffmpeg to fix duration
+        with tempfile.NamedTemporaryFile(suffix=ext) as temp:
+            temp.write(data)
+            temp.flush()
+            with tempfile.NamedTemporaryFile(suffix=ext) as temp_fixed:
+                print(".", end="", flush=True)
+                result = subprocess.run(["ffmpeg", "-y", "-i", temp.name, "-codec", "copy", temp_fixed.name],
+                                        capture_output=True)
+                if result.returncode != 0:
+                    raise RuntimeError(f"Run ffmpeg failed with code {result.returncode}, Error occurred:\n{result.stderr}")
+                temp_fixed.seek(0)
+                data = temp_fixed.read()
+    with tempfile.NamedTemporaryFile(suffix=ext) as temp:
+        temp.write(data)
+        temp.flush()
+        with tempfile.NamedTemporaryFile(suffix=".gif") as gif:
+            from moviepy.editor import VideoFileClip
+            clip = VideoFileClip(temp.name)
+            clip.write_gif(gif.name, logger=None)
+            gif.seek(0)
+            return gif.read()
+
+
+def opermize_gif(data: bytes) -> bytes:
+    with tempfile.NamedTemporaryFile() as gif:
+        gif.write(data)
+        gif.flush()
+        # use gifsicle to optimize gif
+        result = subprocess.run(["gifsicle", "--batch", "--optimize=3", "--colors=256", gif.name],
+                                capture_output=True)
+        if result.returncode != 0:
+            raise RuntimeError(f"Run gifsicle failed with code {result.returncode}, Error occurred:\n{result.stderr}")
+        gif.seek(0)
+        return gif.read()
+
+
+def _convert_image(data: bytes, mimetype: str) -> (bytes, int, int):
+    image: Image.Image = Image.open(BytesIO(data))
     new_file = BytesIO()
-    image.save(new_file, "png")
-    w, h = image.size
+    suffix = mimetypes.guess_extension(mimetype)
+    if suffix:
+        suffix = suffix[1:]
+    # Determine if the image is a GIF
+    is_animated = getattr(image, "is_animated", False)
+    if is_animated:
+        frames = [frame.convert("RGBA") for frame in ImageSequence.Iterator(image)]
+        # Save the new GIF
+        frames[0].save(
+            new_file,
+            format='GIF',
+            save_all=True,
+            append_images=frames[1:],
+            loop=image.info.get('loop', 0),  # Default loop to 0 if not present
+            duration=image.info.get('duration', 100),  # Set a default duration if not present
+            transparency=image.info.get('transparency', 255),  # Default to 255 if transparency is not present
```

Comment:
Can I ask what this is for @xz-dev ?

With this, some of my stickers with a transparent background get a solid black background, some get a solid #00FF00 background. No idea why, but without this, it retains the transparency perfectly.

Reply:
I think it's my mistake.
I also encountered this problem, but I don't have time to debug recently.
I will test to delete it, Thank you

Reply:
Fixed at https://github.com/maunium/stickerpicker/pull/78/commits/a83bc15208f6ddf59b555c8f30c0e6496cda2a89